lib/SIL/IR/TypeLowering.cpp

```diff
@@ -2447,38 +2447,27 @@ static CanAnyFunctionType getStoredPropertyInitializerInterfaceType(
 /// (property-type) -> backing-type.
 static CanAnyFunctionType getPropertyWrapperBackingInitializerInterfaceType(
                                                      TypeConverter &TC,
-                                                     VarDecl *VD) {
+                                                     VarDecl *VD,
+                                                     bool fromWrappedValue) {
   CanType resultType =
       VD->getPropertyWrapperBackingPropertyType()->getCanonicalType();
 
   auto *DC = VD->getInnermostDeclContext();
-  CanType inputType =
-    VD->getPropertyWrapperInitValueInterfaceType()->getCanonicalType();
-
-  auto sig = DC->getGenericSignatureOfContext();
-
-  AnyFunctionType::Param param(
-      inputType, Identifier(),
-      ParameterTypeFlags().withValueOwnership(ValueOwnership::Owned));
-  // FIXME: Verify ExtInfo state is correct, not working by accident.
-  CanAnyFunctionType::ExtInfo info;
-  return CanAnyFunctionType::get(getCanonicalSignatureOrNull(sig), {param},
-                                 resultType, info);
-}
-
-static CanAnyFunctionType getPropertyWrapperInitFromProjectedValueInterfaceType(TypeConverter &TC,
-                                                                                VarDecl *VD) {
-  CanType resultType =
-      VD->getPropertyWrapperBackingPropertyType()->getCanonicalType();
-
-  Type interfaceType = VD->getPropertyWrapperProjectionVar()->getInterfaceType();
-  if (interfaceType->hasArchetype())
-    interfaceType = interfaceType->mapTypeOutOfContext();
+  CanType inputType;
+  if (fromWrappedValue) {
+    inputType = VD->getPropertyWrapperInitValueInterfaceType()->getCanonicalType();
+  } else {
+    Type interfaceType = VD->getPropertyWrapperProjectionVar()->getInterfaceType();
+    if (interfaceType->hasArchetype())
+      interfaceType = interfaceType->mapTypeOutOfContext();
 
-  CanType inputType = interfaceType->getCanonicalType();
+    inputType = interfaceType->getCanonicalType();
+  }
 
-  auto *DC = VD->getInnermostDeclContext();
-  auto sig = DC->getGenericSignatureOfContext();
+  GenericSignature sig;
+  auto *closure = dyn_cast<AbstractClosureExpr>(DC);
+  if (!closure || closure->getCaptureInfo().hasGenericParamCaptures())
```

Comment:
You probably want to use TC.getLoweredLocalCaptures() and not closure->getCaptureInfo(); the former is the transitive closure of the latter.

I believe something like the following will demonstrate the problem:

```
func outer<T>(_: T) {
  func inner() { print(T.self) }

  let closure = { inner() }
  // closure references inner(), which requires the generic context
}
```